enhancements/installer/cluster-api-based-installations.md

```diff
@@ -0,0 +1,347 @@
+- Nutanix: [infrastructure.cluster.x-k8s.io/v1beta1][nutanixCRD]
+- OpenStack: [infrastructure.cluster.x-k8s.io/v1alpha7][openstackCRD]
+- vSphere: [infrastructure.cluster.x-k8s.io/v1beta1][vsphereCRD]
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Overview
+
+In a typical CAPI installation, manifests indicating the desired cluster configuration are applied to a
+management cluster. In order to keep `openshift-install` free of new external runtime dependencies,
+the management-cluster dependencies will be [embedded][embed] into the `openshift-install` binary,
+extracted at runtime, and cleaned up afterward.  This approach is similar to what we have been using
+for Terraform.
+
+In its current configuration, the Installer embds the Terraform and cloud-specific provider binaries
+within the `openshift-install` binary and extracts them at runtime. The Installer produces the Terraform
+configuration files and invokes Terraform using the `tf-exec` library.
+
+![alt terraform diagram](terraform_embedded.jpg)
+
+
+We can follow a similar pattern to run CAPI controllers locally on the Installer host. In addition
+to the CAPI controller binaries, `kube-apiserver` and `etcd` are embedded in order to run a local
+control plane, orchestrated with `envtest`.
+
+![alt capi diagram](capi_embedded.jpg)
+
+#### Local control plane
+
+The local control plane is setup using the previously available work done in Controller Runtime through [envtest][envtest].
+Envtest was born due to a necessity to run integration tests for controllers against a real API server, register webhooks
+(conversion, admission, validation), and managing the lifecycle of Custom Resource Definitions.
+
+Over time, `envtest` matured in a way that now can be used to run controllers in a local environment,
+reducing or eliminating the need for a full Kubernetes cluster to run controllers.
+
+At a high level, the local control plane is responsible for:
+- Setting up certificates for the apiserver and etcd.
+- Running (and cleaning up, on shutdown) the local control plane components.
+- Installing any required component, like Custom Resource Definitions (CRDs)
+    - For Cluster API core the CRDs are stored in `data/data/cluster-api/core-components.yaml`.
+    - Infrastructure providers are expected to store their components in `data/data/cluster-api/<name>-infrastructure-components.yaml`
+- Upon install, the local control plane takes care of modifying any webhook (conversion, admission, validation) to point to the `host:post` combination assigned.
+    - Each controller manager will have its own `host:port` combination assigned.
+    - Certificates are generated and injected in the server, and the client certs in the api-server webhook configuration.
+- For each process that the local control plane manages, a health check (ping to `/healthz`) is required to pass similarly how, when running in a Deployment, a health probe is configured.
```

Comment:
What checks this? Will it be code that we write as part of the installer?

Will  this be checked once as a gate before starting installation (which itself might just mean applying the manifests)? Or continuously? What happens when it does not pass?

Reply:
The check is done both in envtest and similary in the adapted code we have for the installer to run standalone controllers, see https://github.com/openshift/installer/blob/666a25c890f1fe8757e88bfa6d79168656668ca8/pkg/clusterapi/internal/process/process.go#L196-L227

The `process.State` struct described above waits until a predetermined timeout for the controller to be up and running, it mostly expects the controller to respond to the `/healthz` endpoint. Once it's up, it stops checking.

If the endpoint does not pass within timeout the process is closed and returns an error, which is then passed to the installer.